{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#Input: Directory for Spreeadsheet\n",
    "#Input: Spreadsheet of genes, markers\n",
    "#Input: Directory to save files to\n",
    "\n",
    "#For Each guide of input Gene:\n",
    "    #Output: CRISPR Vector File\n",
    "    #Output: Deathstar Ins File\n",
    "    #Output: Primers\n",
    "import sys\n",
    "import csv\n",
    "import re\n",
    "import os\n",
    "from collections import defaultdict\n",
    "\n",
    "# ****Define functions****\n",
    "#Locates gene in chr csv file.\n",
    "#input: gene of interest \"Kpha####\", chr.csv\n",
    "#returns: array[gene_exon_start, gene_exon_end, is_complement]\n",
    "def chr_search(gene, file):\n",
    "    open(file, \"rt\")\n",
    "    with open(file, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        column_1 = [row[\"1\"]for row in reader]\n",
    "        x.close()\n",
    "    with open(file, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        column_2 = [row[\"2\"]for row in reader]\n",
    "        x.close()\n",
    "    with open(file, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        column_3 = [row[\"3\"]for row in reader]\n",
    "        x.close()\n",
    "    with open(file, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        column_4 = [row[\"4\"]for row in reader]\n",
    "        x.close()\n",
    "    with open(file, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        column_5 = [row[\"5\"]for row in reader]\n",
    "        x.close()\n",
    "    with open(file, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        column_6 = [row[\"6\"]for row in reader]\n",
    "        x.close()\n",
    "    with open(file, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        column_7 = [row[\"7\"]for row in reader]\n",
    "        x.close()\n",
    "    with open(file, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        column_8 = [row[\"8\"]for row in reader]\n",
    "        x.close()\n",
    "    with open(file, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        column_9 = [row[\"9\"]for row in reader]\n",
    "        x.close()\n",
    "   \n",
    "    exon_coord_line = \"\"\n",
    "    #now search for the gene and get it's coordinates\n",
    "    for iteration, line in enumerate(column_2):\n",
    "        if (line == \"/label=Kpha\" + i) and (line_last == \"exon\"):\n",
    "            print(line_last)\n",
    "            print(\"Found {} at line {}\".format(gene, iteration))\n",
    "            exon_coord_line = iteration - 1\n",
    "            print(exon_coord_line)\n",
    "            break\n",
    "        \n",
    "        line_last = line\n",
    "\n",
    "    #searching column 3 for coordinate and storing\n",
    "    for iteration, line in enumerate(column_3):\n",
    "        if iteration == exon_coord_line:\n",
    "            print(\"Exon coordinates are {} for {}\".format(line, gene_i))\n",
    "            exon_coords = line\n",
    "\n",
    "    #separate exon_coords string into first number and second number      \n",
    "    print(exon_coords)\n",
    "    gene_exon_start = \"\"\n",
    "    gene_exon_end = \"\"\n",
    "    num_list =[\"0\", \"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"8\", \"9\"]\n",
    "    gene_start_recorded = False\n",
    "    complement = False\n",
    "    for character in exon_coords:\n",
    "        if character == \"c\":\n",
    "            complement = True\n",
    "        if gene_start_recorded == False:\n",
    "            if character in num_list:\n",
    "                gene_exon_start += character\n",
    "            elif (character == \".\"):\n",
    "                gene_start_recorded = True\n",
    "        if gene_start_recorded == True:\n",
    "            if character in num_list:\n",
    "                gene_exon_end += character\n",
    "\n",
    "    #return the coordinates            \n",
    "    print(gene_exon_start)\n",
    "    print(gene_exon_end)\n",
    "    gene_coord_array = [gene_exon_start, gene_exon_end, complement]\n",
    "    return gene_coord_array\n",
    "\n",
    "\n",
    "#gene_start and gene_end should be int vars\n",
    "#complement is boolean\n",
    "#returns: array[locus_seq_fwd, locus_seq_complement]\n",
    "def locus_seq_string(file, gene_start, gene_end, complement):\n",
    "    #store all csv headers as iterable arrays \n",
    "    open(file, \"rt\")\n",
    "    with open(file, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        column_1 = [row[\"1\"]for row in reader]\n",
    "        x.close()\n",
    "    with open(file, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        column_2 = [row[\"2\"]for row in reader]\n",
    "        x.close()\n",
    "    with open(file, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        column_3 = [row[\"3\"]for row in reader]\n",
    "        x.close()\n",
    "    with open(file, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        column_4 = [row[\"4\"]for row in reader]\n",
    "        x.close()\n",
    "    with open(file, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        column_5 = [row[\"5\"]for row in reader]\n",
    "        x.close()\n",
    "    with open(file, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        column_6 = [row[\"6\"]for row in reader]\n",
    "        x.close()\n",
    "    with open(file, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        column_7 = [row[\"7\"]for row in reader]\n",
    "        x.close()\n",
    "    with open(file, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        column_8 = [row[\"8\"]for row in reader]\n",
    "        x.close()\n",
    "    with open(file, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        column_9 = [row[\"9\"]for row in reader]\n",
    "        x.close()\n",
    "    \n",
    "    \n",
    "    #find \"ORIGIN\" in column 1, marking where seuqence starts\n",
    "    print(\"Parsing {} for sequence origin\".format(gene_i_chr))\n",
    "    for iteration, line in enumerate(column_1):\n",
    "        if line == \"ORIGIN\":\n",
    "            seq_begin = iteration + 1\n",
    "        elif line == \"//\":\n",
    "            seq_end = iteration - 1\n",
    "            print(seq_begin)\n",
    "            print(seq_end)\n",
    "    print(\"Found sequence origin coords. Concatenating chromosome as string.\")  \n",
    "    \n",
    "    \n",
    "    chr_seq = \"\"\n",
    "    for index, row in enumerate(column_2):\n",
    "        if ((index >= seq_begin) and (index <= seq_end)):\n",
    "            chr_seq += column_3[index] + column_4[index] + column_5[index] + column_6[index] + column_7[index] + column_8[index]\n",
    "            \n",
    "    print(\"Finished concatenating chromosome sequence\")\n",
    "    \n",
    "    a = 0\n",
    "    for character in chr_seq:\n",
    "        a += 1\n",
    "    \n",
    "    print(\"Chromosome is {} bp long\".format(a))\n",
    "    \n",
    "    print(\"Locating gene locus within chromosome.\")\n",
    "    gene_locus_seq = \"\"\n",
    "    gene_locus_seq_complement = \"\"\n",
    "    gene_locus_seq_fwd = \"\"\n",
    "    n = 1\n",
    "    b = 0\n",
    "    if gene_i_complement == False:\n",
    "        for character in chr_seq:\n",
    "            if (n >= gene_start) and (n <= gene_end):\n",
    "                gene_locus_seq_fwd += character\n",
    "                if character == \"a\":\n",
    "                    gene_locus_seq_complement += \"t\"\n",
    "                elif character == \"t\":\n",
    "                    gene_locus_seq_complement += \"a\"\n",
    "                elif character == \"c\":\n",
    "                    gene_locus_seq_complement += \"g\"\n",
    "                elif character == \"g\":\n",
    "                    gene_locus_seq_complement += \"c\"\n",
    "                b += 1\n",
    "            n += 1\n",
    "        \n",
    "\n",
    "    elif gene_i_complement == True:\n",
    "        for character in chr_seq:\n",
    "            if (n >= gene_start) and (n <= gene_end):\n",
    "                gene_locus_seq_complement += character\n",
    "                if character == \"a\":\n",
    "                    gene_locus_seq_fwd += \"t\"\n",
    "                elif character == \"t\":\n",
    "                    gene_locus_seq_fwd += \"a\"\n",
    "                elif character == \"c\":\n",
    "                    gene_locus_seq_fwd += \"g\"\n",
    "                elif character == \"g\":\n",
    "                    gene_locus_seq_fwd += \"c\"\n",
    "            n += 1\n",
    "        gene_locus_seq_fwd = gene_locus_seq_fwd[::-1]\n",
    "        gene_locus_seq_complement = gene_locus_seq_complement[::-1]\n",
    "\n",
    "    print(\"Locus sequence is {} bp long\".format(b))\n",
    "\n",
    "    #print(chr_seq)\n",
    "    \n",
    "    locus = [gene_locus_seq_fwd, gene_locus_seq_complement]\n",
    "    return locus\n",
    "\n",
    "#****menu prompts****\n",
    "OS = input(\"Unix or Windows OS? (U/W) \")\n",
    "while not (OS == 'U' or OS == 'W'):\n",
    "    print(\"Invalid Response.\")\n",
    "    OS = input(\"Unix or Windows OS? (U/W) \")\n",
    "\n",
    "#obtain the active directory\n",
    "###C:\\Users\\Tim\\Desktop\\Python_Project_Files\n",
    "active_dir = input(\"Please enter the directory for the Python_Project_Files folder: \")\n",
    "while not os.path.exists(active_dir):\n",
    "    if not (active_dir == \"E\" or active_dir == \"e\"):\n",
    "        active_dir = input(\"\"\"The directory entered does not exist, \n",
    "            please re-enter the working directory or enter \"E\" to exit the program: \"\"\")\n",
    "    else: exit()\n",
    "        \n",
    "if (os.path.exists(active_dir)):\n",
    "    print(\"Directory found successfully\")\n",
    "    \n",
    "guide_num_start = input(\"\"\"For R-vector creation, please enter the \n",
    "guide number to start from (INTEGER ONLY, don't include 'gND'): \"\"\")\n",
    "\n",
    "#Windows\n",
    "if (OS == 'W'):\n",
    "    input_folder = (active_dir + \"\\input_files\")\n",
    "    print(\"Given folder directory: \" + input_folder)\n",
    "    while not os.path.exists(input_folder):\n",
    "        if not (active_dir == \"E\" or active_dir == \"e\"):\n",
    "            active_dir = input(\"\"\"The input_folder could not be found in the provided directory\n",
    "            please re-enter the working directory or enter \"E\" to exit the program: \"\"\")\n",
    "            input_folder = (active_dir + \"\\input_files\")\n",
    "        else: exit()\n",
    "    if (os.path.exists(input_folder)):\n",
    "        print(\"Folder found successfully\")\n",
    "        \n",
    "    chr_1 = input_folder + \"\\chr1.csv\"\n",
    "    chr_2 = input_folder + \"\\chr2.csv\"\n",
    "    chr_3 = input_folder + \"\\chr3.csv\"\n",
    "    chr_4 = input_folder + \"\\chr4.csv\"\n",
    "    guide_list = input_folder + \"\\KO_guide_List.csv\"\n",
    "    input_ss = input_folder + \"\\Python_Project_Input.csv\"\n",
    "    \n",
    "#Unix (iOS or Linux)       \n",
    "if (OS == 'U'):\n",
    "    input_folder = (active_dir + \"/input_files\")\n",
    "    print(\"Given folder directory: \" + input_folder)\n",
    "    while not os.path.exists(input_folder):\n",
    "        if not (active_dir == \"E\" or active_dir == \"e\"):\n",
    "            active_dir = input(\"\"\"The input_folder could not be found in the provided directory\n",
    "            please re-enter the working directory or enter \"E\" to exit the program: \"\"\")\n",
    "            input_folder = (active_dir + \"\\input_files\")\n",
    "        else: exit()\n",
    "    if (os.path.exists(input_folder)):\n",
    "        print(\"Folder found successfully\")\n",
    "        \n",
    "    chr_1 = input_folder + \"/chr1.csv\"\n",
    "    chr_2 = input_folder + \"/chr2.csv\"\n",
    "    chr_3 = input_folder + \"/chr3.csv\"\n",
    "    chr_4 = input_folder + \"/chr4.csv\"\n",
    "    guide_list = input_folder + \"/KO_guide_List.csv\"\n",
    "    input_ss = input_folder + \"/Python_Project_Input.csv\"\n",
    "\n",
    "#find the appropriate folders\n",
    "if (os.path.exists(chr_1) and os.path.exists(chr_2) \n",
    "    and os.path.exists(chr_3) and os.path.exists(chr_4)):\n",
    "    print(\"All input files located. Proceeding to parse Python_Project_Input.csv\")\n",
    "else: \n",
    "    print(\"\"\"The files needed to run the program are not present.\n",
    "    The \"input_files\" folder should contain:\n",
    "        chr1.csv,\n",
    "        chr2.csv,\n",
    "        chr3.csv,\n",
    "        chr4.csv,\n",
    "        KO_guide_List.csv,\n",
    "        Python_Project_Input.csv\n",
    "    The program will now close.\"\"\")\n",
    "    exit()\n",
    "    \n",
    "#get input data from Python_Project_Input.csv\n",
    "\n",
    "open(input_ss, \"rt\")\n",
    "with open(input_ss, \"rt\") as i:\n",
    "    reader = csv.DictReader(i)\n",
    "    target_gene = [row[\"target_gene\"]for row in reader]\n",
    "    i.close()\n",
    "\n",
    "\n",
    "with open(input_ss, \"rt\") as i:\n",
    "    reader = csv.DictReader(i)\n",
    "    vec_marker = [row[\"r-vec_marker\"]for row in reader]\n",
    "    i.close()\n",
    "\n",
    "#This for loop will ocmprise the rest of the program\n",
    "#Every action will be performed for every gene in this target_gene array\n",
    "for iteration, i in enumerate(target_gene):\n",
    "    gene_i = \"Kpha\" + i\n",
    "    vec_marker_i = vec_marker[iteration]\n",
    "    print(\"Gene {}:\\t{} with {} marker.\".format((iteration + 1), gene_i, vec_marker[iteration]))\n",
    "    \n",
    "    int_i = int(i)\n",
    "    \n",
    "    if ((int_i >= 1355) and (int_i <= 2978)):\n",
    "        print(\"Gene {} was found in Chromosome 1. Beginning file generation\".format(gene_i))\n",
    "        #parse chr1.csv for gene_id\n",
    "        gene_i_chr = chr_1\n",
    "        gene_coord_output = chr_search(gene_i, gene_i_chr)\n",
    "        \n",
    "    elif ((int_i >= 31) and (int_i <= 1353)):\n",
    "        print(\"Gene {} was found in Chromosome 2. Beginning file generation\".format(gene_i))\n",
    "        gene_i_chr = chr_2\n",
    "        gene_coord_output = chr_search(gene_i, gene_i_chr)\n",
    "        \n",
    "    elif ((int_i >= 3018) and (int_i <= 4303)):\n",
    "        print(\"Gene {} was found in Chromosome 3. Beginning file generation\".format(gene_i))\n",
    "        gene_i_chr = chr_3\n",
    "        gene_coord_output = chr_search(gene_i, gene_i_chr)\n",
    "        \n",
    "    elif ((int_i >= 4305) and (int_i <= 5290)):\n",
    "        print(\"Gene {} was found in Chromosome 4. Beginning file generation\".format(gene_i))\n",
    "        gene_i_chr = chr_4\n",
    "        gene_coord_output = chr_search(gene_i, gene_i_chr)\n",
    "    \n",
    "    print(gene_coord_output)\n",
    "    \n",
    "    for index, var in enumerate(gene_coord_output):\n",
    "        if index == 0:\n",
    "            gene_i_start = var\n",
    "        elif index == 1:\n",
    "            gene_i_end = var\n",
    "        elif index == 2:\n",
    "            gene_i_complement = var\n",
    "    \n",
    "    print(gene_i + \" starts at \" + gene_i_start)\n",
    "    print(gene_i + \" ends at \" + gene_i_end)\n",
    "    print(\"Sequence is complement: \" + str(gene_i_complement))\n",
    "    \n",
    "    gene_i_start_int = int(gene_i_start)\n",
    "    gene_i_end_int = int(gene_i_end)\n",
    "    \n",
    "    #gene string of gene sequence\n",
    "    \n",
    "    ##loci will have + 100bp homology arms in the event that guide is very near the beginning\n",
    "    locus_array = locus_seq_string(gene_i_chr, gene_i_start_int - 100, gene_i_end_int + 100, gene_i_complement)\n",
    "                \n",
    "    for index, var in enumerate(locus_array):\n",
    "        if index == 0:\n",
    "            locus_seq_fwd = var\n",
    "        elif index == 1:\n",
    "            locus_seq_complement = var\n",
    "            \n",
    "    print(locus_seq_fwd)\n",
    "    print(locus_seq_complement)\n",
    "    \n",
    "   \n",
    "    #if no output folder, create one    \n",
    "    #in output folder, save string as locus file\n",
    "    if not (os.path.exists(\"C:\\\\Users\\\\Tim\\\\Desktop\\\\Python_Project_Files\\\\output_files\")):\n",
    "        print(\"Creating output_files folder. Created files will be stored here.\")\n",
    "        os.mkdir(\"C:\\\\Users\\\\Tim\\\\Desktop\\\\Python_Project_Files\\\\output_files\")\n",
    "        os.mkdir(\"C:\\\\Users\\\\Tim\\\\Desktop\\Python_Project_Files\\\\output_files\\\\{}_{}\".format(gene_i, vec_marker_i))\n",
    "    \n",
    "    else:\n",
    "        print(\"output_files folder found. Created files will be stored here.\")\n",
    "        if not (os.path.exists(\"C:\\\\Users\\\\Tim\\\\Desktop\\\\Python_Project_Files\\\\output_files\\\\{}_{}\".format(gene_i, vec_marker_i))):\n",
    "            os.mkdir(\"C:\\\\Users\\\\Tim\\\\Desktop\\\\Python_Project_Files\\\\output_files\\\\{}_{}\".format(gene_i, vec_marker_i))\n",
    "    \n",
    "    #Parse database ss for guide sequences\n",
    "    \n",
    "    #Create CRISPR R-vec plasmid based on guide sequence\n",
    "    \n",
    "    #Find guide seqences in forward or complement string\n",
    "    \n",
    "    #Create deathstar string based on FWD read only, or DS won't be read\n",
    "    \n",
    "    #Write locus seq to Genbank file in output folder\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# C:\\Users\\Tim\\Desktop\\Python_Project_Files\n",
    "import sys\n",
    "import csv\n",
    "import re\n",
    "import os\n",
    "from collections import defaultdict\n",
    "\n",
    "def seq_to_gb_file(seq_start, seq_end, primer_array):\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
